secure-device-storage/sds-android/securedevicestorage/src/main/java/de/adorsys/android/securedevicestorage/SecurePreferences.java

```diff
@@ -24,52 +24,94 @@
-                KeystoreTool.generateKeyPair(context);
+    public static void setValue(@NonNull String key, @NonNull String value,
+                                @NonNull Context context,
+                                @NonNull SecureMethod secureMethod) {
+        if (secureMethod.equals(SecureMethod.METHOD_ENCRYPT)) {
+            try {
+                if (!KeystoreTool.keyPairExists()) {
+                    KeystoreTool.generateKeyPair(context);
+                }
+            } catch (CertificateException | NoSuchAlgorithmException | IOException
+                    | KeyStoreException | UnrecoverableKeyException
+                    | InvalidAlgorithmParameterException | NoSuchProviderException e) {
+                Log.e(SecurePreferences.class.getName(), e.getMessage(), e);
+                return;
             }
-        } catch (CertificateException | NoSuchAlgorithmException | IOException
-                | KeyStoreException | UnrecoverableKeyException
-                | InvalidAlgorithmParameterException | NoSuchProviderException e) {
-            Log.e(SecurePreferences.class.getName(), e.getMessage(), e);
-            return;
-        }
 
-        String transformedValue = null;
-        try {
-            transformedValue = KeystoreTool.encryptMessage(context, value);
-        } catch (NoSuchPaddingException | NoSuchAlgorithmException | NoSuchProviderException
-                | IOException | UnrecoverableEntryException | KeyStoreException
-                | InvalidKeyException | CertificateException e) {
+            String transformedValue = null;
+            try {
+                transformedValue = KeystoreTool.encryptMessage(context, value);
+            } catch (NoSuchPaddingException | NoSuchAlgorithmException | NoSuchProviderException
+                    | IOException | UnrecoverableEntryException | KeyStoreException
+                    | InvalidKeyException | CertificateException e) {
 
-            Log.e(SecurePreferences.class.getName(), e.getMessage(), e);
-        }
-        if (transformedValue != null) {
-            setSecureValue(key, transformedValue, context);
+                Log.e(SecurePreferences.class.getName(), e.getMessage(), e);
+            }
+            if (transformedValue != null) {
+                setSecureValue(key, transformedValue, context);
+            } else {
+                Log.e(SecurePreferences.class.getName(),
+                        context.getString(R.string.message_problem_encryption));
+            }
         } else {
-            Log.e(SecurePreferences.class.getName(),
-                    context.getString(R.string.message_problem_encryption));
+            byte[] saltBytes = KeystoreTool.calculateSalt();
+            String salt;
+            try {
+                if (saltBytes != null) {
+                    salt = new String(saltBytes, 0, saltBytes.length, KEY_CHARSET);
+                } else {
+                    Log.e(SecurePreferences.class.getName(),
```

Comment:
I think it's a good idea to throw a custom RuntimeException here so that the developer can get feedback that it didn't work